# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/gc/GCParallelTask.cpp
# Commit: 28829816a629
# Full Hash: 28829816a629b28125db3b4c65c5e22518e167e8
# Author: Jon Coppeard <jcoppeard@mozilla.com>
# Date: 2022-03-09 21:37:17
# Regressor Bug: 1643586
# File Overlap Count: 1
# Description:
#   Bug 1643586 - Part 6: Stop using JSContext for GC helper thread tasks r=jandem
#   
#   This patch moves the debug-only GC state from JSContext to JSFreeOp. Support
#   functions for this move to FreeOp.h and most RAII classes move to GCInternals.h.
#   
# ==============================================================================

diff -r 410d1b57bb09 -r 28829816a629 js/src/gc/GCParallelTask.cpp
--- a/js/src/gc/GCParallelTask.cpp	Wed Mar 09 10:17:07 2022 +0000
+++ b/js/src/gc/GCParallelTask.cpp	Wed Mar 09 10:17:08 2022 +0000
@@ -9,6 +9,8 @@
 #include "mozilla/MathAlgorithms.h"
 #include "mozilla/Maybe.h"
 
+#include "gc/FreeOp.h"
+#include "gc/GCInternals.h"
 #include "gc/ParallelWork.h"
 #include "vm/HelperThreadState.h"
 #include "vm/Runtime.h"
@@ -156,18 +158,22 @@
 
   setRunning(lock);
 
-  JS::ContextOptions options;  // The GC doesn't use any context options.
-  AutoSetHelperThreadContext usesContext(options, lock);
-  AutoSetContextRuntime ascr(gc->rt);
-  gc::AutoSetThreadIsPerformingGC performingGC;
+  JSFreeOp freeOp(gc->rt, false);
+  MOZ_RELEASE_ASSERT(TlsFreeOp.init());
+  TlsFreeOp.set(&freeOp);
+
   runTask(lock);
 
+  TlsFreeOp.set(nullptr);
+
   setFinished(lock);
 }
 
 void GCParallelTask::runTask(AutoLockHelperThreadState& lock) {
   // Run the task from either the main thread or a helper thread.
 
+  gc::AutoSetThreadIsPerformingGC performingGC;
+
   // The hazard analysis can't tell what the call to func_ will do but it's not
   // allowed to GC.
   JS::AutoSuppressGCAnalysis nogc;